### ISSUE TITLE
Feature request : add weight normalization

### ISSUE THREAD

**Issue by taki0112** [tensorflow/tensorflow#14070]
can you implement [weight norm](https://arxiv.org/pdf/1602.07868.pdf) ?

I want to use it as follows.
```python
    x = tf.layers.conv2d(x, filter_size=32, kernel_size=[3,3], strides=2)
    x = weight_norm(x)
```
Is it possible?

**Comment by jart** [comment]
@martinwicke do you have thoughts on this [weight normalization](https://arxiv.org/pdf/1602.07868.pdf) paper?

**Comment by martinwicke** [comment]
It is definitely possible, and contributions implementing it are welcome.

**Comment by facaiy** [comment]
@taki0112 Not sure whether I understand weight normalization correctly?

It is indeed `Dense` layer which always normalizes its kernel when called (and add a new parameter `g`), right?

<img width="298" alt="image" src="https://user-images.githubusercontent.com/1112263/32213990-61f1d676-be58-11e7-8469-dab62722a46e.png">

**Comment by taki0112** [comment]
@facaiy 
Yes, i think may be right !

**Comment by divyanshj16** [comment]
@martinwicke 
I want to take this! Can somebody help to start? This will be my first contribution!

**Comment by taki0112** [comment]
@divyanshj16 
oh.. really thank you
It would be great if we could see it as soon as possible.
Thank you for your implementation.

**Comment by King-Of-Knights** [comment]
@taki0112 , I found a relative code here, can some one verify it?
[https://github.com/PFCM/weightnorm/blob/master/weightnorm/weightnorm.py]

**Comment by yaroslavvb** [comment]
PS, for details on how to derive the formulas from scratch, some discussion is here -- https://discuss.openai.com/t/proof-of-weight-normalizations-mechanism/3026/3

**Comment by llan-ml** [comment]
Any updates? Can we add the weight normalization in the `self.build()` method of `tf.Layer`.
For example, in the scenario of `Dense` layer, we can add weight normalization by modifying https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/layers/core.py#L152:44 as follows:
<pre>
kernel = self.add_variable("kernel", ...)
if self._weight_norm:
    g = self.add_variable("wn/g", ...)
    self.kernel = nn_impl.l2_normalize(kernel, axis=0) * g
else:
    self.kernel = kernel
</pre>
Similar modifications could be also made for `tensorflow.python.layers.convolutional._Conv`.
Do you have any suggestions?

**Comment by seanpmorgan** [comment]
This has been implemented in tensorflow/addons. Issue can be closed.

[Code](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/wrappers.py)
[Example](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/examples/layers_weightnormalization.ipynb)

**Comment by facaiy** [comment]
Thanks, Sean.
